toolboxes/mri_core/mri_core_girf_correction.cpp

```diff
@@ -1,160 +1,282 @@
+                        }
+                    }
+                    std::rotate(rotated.begin() + k * rotated.get_size(0), rotated.begin() + k * rotated.get_size(0) + (nlines / 2 + nlines_new / 2),
+                                rotated.begin() + (k + 1) * rotated.get_size(0));
+                }
+
+                temp = real(permute(rotated, {0, 1}));
+                Gadgetron::maxValue(hoNDArray<float>(temp(slice, 0)), maxTx);
+                Gadgetron::minValue(hoNDArray<float>(temp(slice, 0)), minTx);
 
+                if (maxTx > 100 || minTx < -100)
+                    GERROR("What the heck");
 
+                hoNDFFT<float>::instance()->fftshift1D(rotated);
                 hoNDFFT<float>::instance()->fft1(rotated);
                 hoNDFFT<float>::instance()->fftshift1D(rotated);
-//            auto padded_kernel = std::move(*pad<std::complex<float>,1>(uint64d1(nlines_new), &girf_kernel));
-
-//            rotated *= padded_kernel;
-            rotated *= girf_kernel;
-            float ADCShift = -TE-0.5*gradient_sampling_time;
-                float frequency_resolution = 1.0 / (girf_sampling_time * nlines_new);
-            for (size_t k = 0; k < 3; k++) {
-                for (size_t i = 0; i < nlines_new; i++){
-                    rotated(i,k) *= std::exp(-2if*ADCShift*PI*frequency_resolution*float(i));
+
+                rotated *= padded_kernel;
+
+                float ADCShift = 1.0*(TE + 0.5 * gradient_sampling_time); //Added -ve sign inspired by NAM
+                float frequency_resolution = 1.0 / (gradient_sampling_time * nlines_new);
+                for (size_t k = 0; k < 3; k++)
+                {
+                    for (size_t i = 0; i < nlines_new; i++)
+                    {
+                        // Removed -ve sign needs more testing but was trying to match MATLAB implementation
+                        rotated(i, k) = rotated(i, k) * std::exp(2if * ADCShift * PI * frequency_resolution * float(i));
+                    }
                 }
-            }
 
                 size_t full_domain = size_t(std::round(1.0 / (2 * frequency_resolution * gradient_sampling_time))) * 2;
 
-//                rotated = *pad<std::complex<float>, 1>(uint64d1(full_domain), &rotated);
+                // rotated = *pad<std::complex<float>, 1>(uint64d1(full_domain), &rotated);
 
                 hoNDFFT<float>::instance()->ifftshift1D(rotated);
                 hoNDFFT<float>::instance()->ifft1(rotated);
+                hoNDFFT<float>::instance()->ifftshift1D(rotated);
+
+                temp = real(permute(rotated, {0, 1}));
+                Gadgetron::maxValue(hoNDArray<float>(temp(slice, 0)), maxTx);
+                Gadgetron::minValue(hoNDArray<float>(temp(slice, 0)), minTx);
 
+                if (maxTx > 100 || minTx < -100)
+                    GERROR("What the heck");
 
-//            hoNDFFT<float>::instance()->ifft(&rotated,0);
+                for (size_t k = 0; k < 3; k++)
+                {
+
```

Comment:
What exactly are you trying to do here?

Reply:
There is some magical filtration that is needed to make GIRF work - Its just applying a filter to smooth the zero padding for the gradient waveform and rotating the waveform to the center before cetered fft. I spent some time trying to tweak this and settled on this implementation - there maybe a better way to do this but we know this works and it replicates the matlab implementation.